contrib/mesos/pkg/executor/service/podsource/podsource.go

```diff
@@ -0,0 +1,201 @@
+func Mesos(
+	stop <-chan struct{},
+	out chan<- interface{},
+	podWatch *cache.ListWatch,
+	registry executor.Registry,
+	options ...Option,
+) {
+	source := &Source{
+		stop: stop,
+		out:  out,
+		filters: []Filter{
+			FilterFunc(filterMirrorPod),
+			&registeredPodFilter{registry: registry},
+		},
+	}
+	// note: any filters added by options should be applied after the defaults
+	for _, opt := range options {
+		opt(source)
+	}
+	// reflect changes from the watch into a chan, filtered to include only mirror pods
+	// (have an ConfigMirrorAnnotationKey attr)
+	cache.NewReflector(
+		podWatch,
+		&api.Pod{},
+		cache.NewUndeltaStore(source.send, cache.MetaNamespaceKeyFunc),
+		0,
+	).RunUntil(stop)
+}
+
+func filterMirrorPod(p **api.Pod) (ok bool) {
+	_, ok = (*p).Annotations[kubetypes.ConfigMirrorAnnotationKey]
+	return
+}
+
+type registeredPodFilter struct {
+	priorPodNames, podNames map[podName]string // maps a podName to a taskID
+	registry                executor.Registry
+}
+
+func (rpf *registeredPodFilter) Before(podCount int) {
+	rpf.priorPodNames = rpf.podNames
+	rpf.podNames = make(map[podName]string, podCount)
+}
+
+func (rpf *registeredPodFilter) After() {
+	// detect when pods are deleted and notify the registry
+	for k, taskID := range rpf.priorPodNames {
+		if _, found := rpf.podNames[k]; !found {
+			rpf.registry.Remove(taskID)
+		}
+	}
+}
+
+func (rpf *registeredPodFilter) Accept(p **api.Pod) (ok bool) {
+	if rpod, err := rpf.registry.Update(*p); err == nil {
+		// pod is bound to a task, and the update is compatible
+		// so we'll allow it through
+		*p = rpod.Pod() // use the (possibly) updated pod spec!
+		rpf.podNames[podName{(*p).Namespace, (*p).Name}] = rpod.Task()
+		ok = true
+	} else if rpod != nil {
+		// we were able to ID the pod but the update still failed...
+		log.Warningf("failed to update registry for task %v pod %v/%v: %v",
+			rpod.Task(), (*p).Namespace, (*p).Name, err)
+	}
+	return
+}
+
+// send is an update callback invoked by NewUndeltaStore; it applies all of source.filters
+// to the incoming pod snapshot and forwards a PodUpdate that contains a snapshot of all
+// the pods that were accepted by the filters.
+func (source *Source) send(objs []interface{}) {
+	var (
+		podCount = len(objs)
+		pods     = make([]*api.Pod, 0, podCount)
+	)
+
+	for _, f := range source.filters {
+		f.Before(podCount)
+	}
+foreachPod:
+	for _, o := range objs {
+		p := o.(*api.Pod)
+		for _, f := range source.filters {
+			if f.Accept(&p) {
+				pods = append(pods, p)
+				continue foreachPod
+			}
+		}
+		// unrecognized pod
+		log.V(2).Infof("skipping pod %v/%v", p.Namespace, p.Name)
+	}
+	// TODO(jdef) should these be applied in reverse order instead?
+	for _, f := range source.filters {
+		f.After()
+	}
+
+	u := kubetypes.PodUpdate{
+		Op:     kubetypes.SET,
+		Pods:   pods,
+		Source: MesosSource,
+	}
+	select {
+	case <-source.stop:
+	default:
+		select {
+		case <-source.stop:
```

Comment:
I think this is a personal anti-pattern that sneaks in when I feel paranoid about how Go handles tie breaking in select. This really isn't a solution for it. Should be simplified - thanks for calling it ou.